File: 15_2.ts
import { open } from 'node:fs/promises';

type Interval = [number, number];
type Point = [number, number];

function union(intervals: Interval[]): Interval[] {
  const temp = [...intervals];
  temp.sort(([l0,], [l1,]) => l0 - l1);
  const result: Interval[] = [];
  for (const [l, r] of temp) {
    if (result.length === 0) {
      result.push([l, r]);
    } else {
      const [prevL, prevR] = result[result.length - 1];
      if (prevR + 1 >= l) {
        result[result.length - 1] = [prevL, Math.max(r, prevR)];
      } else {
        result.push([l, r]);
      }
    }
  }
  return result;
}

// const inputFile = '15_test.txt';
// const maxX = 20;
// const maxY = 20;
const inputFile = '15.txt';
const maxX = 4000000;
const maxY = 4000000;

const inputRe = /Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)/;
const finput = await open(inputFile);
const sensors: [Point, number][] = [];
for await (const line of finput.readLines()) {
  const match = line.match(inputRe)!;
  const [sx, sy, bx, by] = [match[1], match[2], match[3], match[4]].map(Number);
  const radius = Math.abs(bx - sx) + Math.abs(by - sy);
  sensors.push([[sx, sy], radius]);
}

const sTime = Date.now();

for (let targetY = 0; targetY <= maxY; ++targetY) {
  const intervals: Interval[] = [];
  for (const [[sx, sy], radius] of sensors) {
    const radiusAtTargetRow = radius - Math.abs(sy - targetY);
    if (radiusAtTargetRow >= 0) {
      const interval: Interval = [sx - radiusAtTargetRow, sx + radiusAtTargetRow];
      intervals.push(interval);
    }
  }
  const intervalUnion = union(intervals);
  let targetX = 0;
  for (let i = 0; i < intervalUnion.length - 1; ++i) {
    const r0 = intervalUnion[i][1];
    const l1 = intervalUnion[i + 1][0];
    if (l1 - r0 === 2) {
      targetX = r0 + 1;
      if (targetX >= 0 && targetX <= maxX) {
        console.log(targetX, targetY, targetX * 4000000 + targetY);
      }
    }
  }
}

console.log(`Time: ${Date.now() - sTime}`);

File: 15_2.cs
using System.Text.RegularExpressions;

namespace App
{

    struct Sensor
    {
        public int x;
        public int y;
        public int radius;
    }


    class Program
    {
        private static List<(int, int)> union(List<(int, int)> intervals)
        {
            var sortedIntervals = new List<(int, int)>(intervals);
            sortedIntervals.Sort();
            var result = new List<(int, int)>();
            foreach (var (l, r) in sortedIntervals)
            {
                if (result.Count == 0)
                {
                    result.Add((l, r));
                }
                else
                {
                    var (prevL, prevR) = result.Last();
                    if (prevR + 1 >= l)
                    {
                        result[result.Count - 1] = (prevL, Math.Max(r, prevR));
                    }
                    else
                    {
                        result.Add((l, r));
                    }
                }
            }
            return result;
        }
        static void Main()
        {
            //const string inputFile = "../../../15_test.txt";
            //const int maxX = 20;
            //const int maxY = 20;
            const string inputFile = "../../../15.txt";
            const int maxX = 4000000;
            const int maxY = 4000000;

            var inputRe = new Regex(@"Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)");
            var sensors = new List<Sensor>();
            foreach (var line in File.ReadLines(inputFile))
            {
                var match = inputRe.Match(line);
                var nums = match.Groups.Values.ToArray()[1..5].Select(g => int.Parse(g.Value)).ToList();
                var (sx, sy, bx, by) = (nums[0], nums[1], nums[2], nums[3]);
                var radius = Math.Abs(bx - sx) + Math.Abs(by - sy);
                sensors.Add(new Sensor { x = sx, y = sy, radius = radius });
            }

            var sTime = DateTime.Now;

            for (var targetY = 0; targetY <= maxY; ++targetY)
            {
                var intervals = new List<(int, int)>();
                foreach (var s in sensors)
                {
                    var radiusAtTargetY = s.radius - Math.Abs(s.y - targetY);
                    if (radiusAtTargetY >= 0)
                    {
                        intervals.Add((s.x - radiusAtTargetY, s.x + radiusAtTargetY));
                    }
                }
                var intervalUnion = union(intervals);
                for (var i = 0; i < intervalUnion.Count - 1; ++i)
                {
                    var (_, r0) = intervalUnion[i];
                    var (l1, _) = intervalUnion[i + 1];
                    if (l1 - r0 == 2)
                    {
                        var targetX = r0 + 1;
                        if (targetX >= 0 && targetX <= maxX)
                        {
                            Console.WriteLine("{0} {1} {2}", targetX, targetY, (long)targetX * 4000000 + targetY);
                        }
                    }
                }
            }
            Console.WriteLine("Time: {0}", DateTime.Now - sTime);
        }
    }
}